File: oc/YeeSigner/YeeSigner/YeeSigner/KeyPair.m
//
//  KeyPair.m
//  YeeSigner
//
//  Created by GB on 2020/6/20.
//  Copyright © 2020 io.yeeco. All rights reserved.
//

#import "KeyPair.h"
#import "libyee_signer.h"
#import "ErrorUtils.h"

@interface KeyPair ()

@property (nonatomic) unsigned int* pointer;

@end

@implementation KeyPair

const unsigned int MINI_SECRET_KEY_LENGTH = 32;
const unsigned int SECRET_KEY_LENGTH = 64;
const unsigned int PUBLIC_KEY_LENGTH = 32;
const unsigned int SIGNATURE_LENGTH = 64;

+ (KeyPair *) fromMiniSecretKey:(NSData* ) miniSecretKey error:(NSError **)error{
        
    unsigned int err = 0;
    unsigned int* pointer = yee_signer_key_pair_from_mini_secret_key(miniSecretKey.bytes, (unsigned int) miniSecretKey.length, &err);
    if(err > 0) {
        *error = [ErrorUtils error:err];
        return nil;
    }
    
    KeyPair* keyPair = [KeyPair alloc];
    keyPair.pointer = pointer;
    return keyPair;
}

+ (KeyPair *) fromSecretKey:(NSData* ) secretKey error:(NSError **)error{
        
    unsigned int err = 0;
    unsigned int* pointer = yee_signer_key_pair_from_secret_key(secretKey.bytes, (unsigned int) secretKey.length, &err);
    if(err > 0) {
        *error = [ErrorUtils error:err];
        return nil;
    }
    
    KeyPair* keyPair = [KeyPair alloc];
    keyPair.pointer = pointer;
    return keyPair;
}

- (NSData *) publicKey:(NSError **)error {
    
    unsigned int err = 0;
    unsigned char result[PUBLIC_KEY_LENGTH];
    yee_signer_public_key(self.pointer, result, PUBLIC_KEY_LENGTH, &err);
    if(err > 0) {
        *error = [ErrorUtils error:err];
        return nil;
    }
    
    NSData* data = [NSData dataWithBytes:(const void *)result length:PUBLIC_KEY_LENGTH];
    return data;
}

- (NSData *) secretKey:(NSError **)error {
    
    unsigned int err = 0;
    unsigned char result[SECRET_KEY_LENGTH];
    yee_signer_secret_key(self.pointer, result, SECRET_KEY_LENGTH, &err);
    if(err > 0) {
        *error = [ErrorUtils error:err];
        return nil;
    }
    
    NSData* data = [NSData dataWithBytes:(const void *)result length:SECRET_KEY_LENGTH];
    return data;
}

- (NSData *) sign: (NSData *)message ctx: (NSData *)ctx error:(NSError **)error{
    
    unsigned int err = 0;
    unsigned char result[SIGNATURE_LENGTH];
    yee_signer_sign(self.pointer, message.bytes, (unsigned int)message.length,
                    result, SIGNATURE_LENGTH, ctx.bytes, (unsigned int)ctx.length, &err);
    if(err > 0) {
        *error = [ErrorUtils error:err];
        return nil;
    }
    
    NSData* data = [NSData dataWithBytes:(const void *)result length:SIGNATURE_LENGTH];
    return data;
}

- (void) free:(NSError **)error{
    
    unsigned int err = 0;
    yee_signer_verifier_free(self.pointer, &err);
    if(err > 0) {
        *error = [ErrorUtils error:err];
    }
}

@end

File: java/src/main/java/io/yeeco/yeesigner/KeyPair.java
package io.yeeco.yeesigner;

public class KeyPair {

    public final static int PUBLIC_KEY_LENGTH = 32;
    public final static int SECRET_KEY_LENGTH = 64;
    public final static int MINI_SECRET_KEY_LENGTH = 32;
    public final static int SIGNATURE_KEY_LENGTH = 64;

    private long pointer;

    public static KeyPair fromMiniSecretKey(byte[] miniSecretKey) throws SignerException {

        byte[] error = new byte[1];
        long pointer = JNI.keyPairFromMiniSecretKey(miniSecretKey, error);
        ErrorUtils.checkErrorCode(error[0]);

        KeyPair instance = new KeyPair();
        instance.pointer = pointer;
        return instance;
    }

    public static KeyPair fromSecretKey(byte[] secretKey) throws SignerException {
        byte[] error = new byte[1];
        long pointer = JNI.keyPairFromSecretKey(secretKey, error);
        ErrorUtils.checkErrorCode(error[0]);

        KeyPair instance = new KeyPair();
        instance.pointer = pointer;
        return instance;
    }

    public byte[] getPublicKey() throws SignerException {

        byte[] publicKey = new byte[PUBLIC_KEY_LENGTH];
        byte[] error = new byte[1];
        JNI.publicKey(pointer, publicKey, error);
        ErrorUtils.checkErrorCode(error[0]);

        return publicKey;
    }

    public byte[] getSecretKey() throws SignerException {
        byte[] secretKey = new byte[SECRET_KEY_LENGTH];
        byte[] error = new byte[1];
        JNI.secretKey(pointer, secretKey, error);
        ErrorUtils.checkErrorCode(error[0]);

        return secretKey;
    }

    public byte[] sign(byte[] message, byte[] ctx) throws SignerException {
        byte[] signature = new byte[SIGNATURE_KEY_LENGTH];
        byte[] error = new byte[1];
        JNI.sign(pointer, message, signature, ctx, error);
        ErrorUtils.checkErrorCode(error[0]);

        return signature;
    }

    @Override
    protected void finalize(){
        byte[] error = new byte[1];
        JNI.keyPairFree(pointer, error);
    }

}
